### PR TITLE
Cirrus: Update CI VM Image to F38

### PR DESCRIPTION
**Pull request by cevich**
Bump F36 -> F37, and F37 -> F38

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by cevich** [comment]
@mtrmac IIRC you know a lot about these things.  Is it okay for me to replace these calls [as suggested in the linter output](https://cirrus-ci.com/task/4748928504037376)?

**Comment by mtrmac** [comment]
I think the answer to that changes a bit based on whether we _require_ Go 1.20 for Podman (https://go.dev/doc/go1.20 , random seed by default).

> libpod/runtime.go:125:2

Would not work, vendor/github.com/docker/docker/pkg/namesgenerator/names-generator.go uses the global RNG. With Go 1.20, we can just delete the seeding to silence the warning without a behavior change.

Ideally, the generator should take an explicit RNG parameter, and Podman should call `NewRand` to maintain one.

> libpod/runtime_test.go:25:2

Supposedly the `Seed` call is an essential part of the test … except that I’m not at all sure this is testing what it is saying it’s testing (and if it _is_ testing that, it should fail with Go < 1.20 AFAICS).

> test/e2e/common_test.go:433:2

Sure, use a `NewRand` locally for that function.

**Comment by cevich** [comment]
Thanks for your input @mtrmac

**Comment by cevich** [comment]
@mtrmac PTAL 1521bd7 see if I got that right.  Are you at all concerned that these changes will be invalid for <F38 which isn't using golang 1.20?

File: libpod/runtime_test.go [hunk 1]
```diff
@@ -19,10 +18,8 @@ func Test_generateName(t *testing.T) {
 	}
 
 	// Test that (*Runtime).generateName returns different names
-	// if called twice, even if the global RNG has the default
-	// seed.
+	// if called twice.
 	n1, _ := r.generateName()
-	rand.Seed(1)
```
Comment (mtrmac) [inline libpod/runtime_test.go:25]: I am not sure that the old test worked as intended, but the new one would really not be testing the “default seed” behavior.

(It seems to me that the concern is that two Podman processes don’t just start generating the same names, significantly raising the risk of collisions. Hence the test that names generated “in default conditions” don’t match the ones with `Seed(1)`.)

File: test/e2e/common_test.go [hunk 2]
```diff
@@ -430,11 +430,11 @@ func GetPortLock(port string) *lockfile.LockFile {
...
 	// To avoid IP collisions of initialize random seed for random IP addresses
-	rand.Seed(time.Now().UnixNano())
+	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
 	// Add GinkgoParallelProcess() on top of the IP address
 	// in case of the same random seed
-	ip3 := strconv.Itoa(rand.Intn(230) + GinkgoParallelProcess())
-	ip4 := strconv.Itoa(rand.Intn(230) + GinkgoParallelProcess())
+	ip3 := strconv.Itoa(rng.Intn(230) + GinkgoParallelProcess())
+	ip4 := strconv.Itoa(rng.Intn(230) + GinkgoParallelProcess())
 	return "10.88." + ip3 + "." + ip4
```
Comment (mtrmac) [inline test/e2e/common_test.go:438]: This looks right.

File: libpod/runtime.go [hunk 3]
```diff
@@ -118,13 +117,6 @@ type Runtime struct {
 	secretsManager *secrets.SecretsManager
 }
 
-func init() {
-	// generateName calls namesgenerator.GetRandomName which the
-	// global RNG from math/rand. Seed it here to make sure we
-	// don't get the same name every time.
-	rand.Seed(time.Now().UnixNano())
-}
-
```
Comment (mtrmac) [inline libpod/runtime.go:127]: Do we now require Go 1.20?  At least `go.mod` says 1.18, I’m not sure what is the authoritative source for that decision.

Without Go 1.20, this could lead to much more frequent collisions, we don’t want that.

**Review by mtrmac — COMMENTED** [review]
I’m worried that fixing the `libpod` part completely cleanly might require changes to `docker/docker/pkg/namesgenerator`.

**Comment by cevich** [comment]
@vrothberg & @mheon Might could use your input on this `rand` situation (in `libpod/runtime.go`).  Seems like breaking-behavior change between golang 1.19 (F37) and 1.20 (F38 + rawhide): i.e. In =<1.19, global RNG seed is initialized at startup vs =>1.20 it's automatically initialized.

File: libpod/runtime.go (hunk 3 above)
Comment (cevich) [reply libpod/runtime.go:127]: We don't require it yet, but won't go use the behavior of the compiler-version?  Or does it stick to what's in `go.mod`?

Comment (mtrmac) [reply libpod/runtime.go:127]: I didn’t actually test that, but I think in _almost all_ cases it uses the compiler’s version, without special casing the `go.mod` declaration.

Which means that if this successfully compiles with Go 1.19, it is going to be seeded with 1 and not work the way we need.

---

One possible short-term workaround would be to just silence the linter; seeding the global RNG doesn’t exactly _hurt_.

It’s not nearly as clean as modifying the names generator to add a RNG parameter… and I’ll let actual Podman maintainer decide whether the silencing is good enough.

File: libpod/runtime_test.go (hunk 1 above)
Comment (mtrmac) [reply libpod/runtime_test.go:25]: I think that depends on whether we end up parametrizing the name generator with a RNG; if we do, we can have a good test.

If not, the version in this PR is probably as good as it gets, I agree.

File: libpod/runtime.go (hunk 3 above)
Comment (cevich) [reply libpod/runtime.go:127]: Is there some kind of compiler tag or fancy wizardry that could be used to include `rand.Seed(time.Now().UnixNano())` **only** when compiling with go 1.19 and under?

> it is going to be seeded with 1 and not work the way we need.

Another idea, maybe we could have podman set `$GODEBUG=randautoseed=0` to force the automatic-behavior to do things the "old" way.

Comment (mtrmac) [reply libpod/runtime.go:127]: > Is there some kind of compiler tag or fancy wizardry that could be used to include `rand.Seed(time.Now().UnixNano())` **only** when compiling with go 1.19 and under?

Oh, yes. Great point, I have completely forgotten about that.

“Build tags”, compare https://github.com/containers/storage/blob/main/pkg/archive/archive_19.go / https://github.com/containers/storage/blob/main/pkg/archive/archive_110.go .  (The old one might then still need a lint suppression comment, but it would be clearly justified.)

Comment (cevich) [reply libpod/runtime.go:127]: so IIUC, this would look something like:

`libpod/runtime_go_120.go`:
```
//go:build !go1.20
// +build !go1.20

package libpod

func init() {
    rand.Seed(time.Now().UnixNano())
}
```

Would that mean the file is skipped for all versions > 1.20 as well?

**Comment by cevich** [comment]
Force-push: Add go <1.20 build tag to handle the Seed() change.  Also added new CI validation task for rawhide - it would have caught this problem earlier.

File: libpod/runtime_pre_go1.10.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
+// In go 1.20 and later, the global RNG is automatically initialized.
```
Comment (mtrmac) [inline libpod/runtime_pre_go1.10.go:1]: Rename from `1.10` to `1.20`?

**Review by mtrmac — COMMENTED** [review]
LGTM. Changing the generator would be a bit cleaner, but also feels rather out of scope.

**Comment by mtrmac** [comment]
> LGTM.

The RNG changes at least, that is. I didn’t check what else is changing, and if anything is failing.

File: libpod/runtime_pre_go1.10.go (hunk 4 above)
Comment (cevich) [reply libpod/runtime_pre_go1.10.go:1]: oh woops.

**Comment by cevich** [comment]
@vrothberg & @mheon un-Ping.  I think Miloslav and I have it worked out.

**Comment by cevich** [comment]
Force-push: Added a simple e2e test to confirm running two containers w/ generated names works.

File: test/e2e/run_test.go [hunk 5]
```diff
@@ -48,6 +48,26 @@ var _ = Describe("Podman run", func() {
 		Expect(session).Should(Exit(0))
 	})
 
+	It("podman run two containers choose different names", func() {
+		session1 := podmanTest.Podman([]string{"run", "-d", ALPINE, "sleep 1m"})
+		session2 := podmanTest.Podman([]string{"run", "-d", ALPINE, "sleep 1m"})
```
Comment (edsantiago) [inline test/e2e/run_test.go:53]: This can't possibly work, and it's going to fail CI with something like "no such executable, 'sleep 1m'". Needs to be `"sleep", "1m"` or something like that, separate tokens.

File: test/e2e/run_test.go [hunk 6]
```diff
@@ -48,6 +48,26 @@ var _ = Describe("Podman run", func() {
 		Expect(session).Should(Exit(0))
 	})
 
+	It("podman run two containers choose different names", func() {
```
Comment (edsantiago) [inline test/e2e/run_test.go:51]: Misleading: should be **container IDs**, not names. The output of `run -d` is a CID.

Unless you really want to test container **names**, in which case you'll need to `podman ps` or `podman inspect` or something else. 

But I'm not sure there's any point to this test: it seems impossible for two containers to have the same CID or name.

**Review by edsantiago — COMMENTED** [review]
I would suggest just removing the new test

File: test/e2e/run_test.go (hunk 6 above)
Comment (mtrmac) [reply test/e2e/run_test.go:51]: https://github.com/containers/podman/blob/0d171e9b7a9857e4809037c152ed1c617e5b5a79/libpod/container_internal.go#L510-L514 suggests that this is names.

---

> It seems impossible for two containers to have the same CID or name.

So I think a test could create a container, record the name, delete it; create another container, record the name, delete it. Then check that the two generated names are different, i.e. that we don’t always generate the same (sequence of) names.

That would actually, unlike the previous unit test ( https://github.com/containers/podman/pull/18326#discussion_r1176766233 ), be a representative test of the whole-process RNG seeding behavior.

OTOH, it would also be an inherently flaky test, possibly generating the same name in the two runs purely by chance of how the RNG is seeded.

Comment (edsantiago) [reply test/e2e/run_test.go:51]: >suggests that this is names

I'm very confused. I see the test running `podman run -d`, which AFAIK only emits CIDs and has only ever emitted CIDs, as has docker. I don't recall ever seeing a name.

But anyhow.  I can't get podman to reuse names even if I hardcode the seed:
```diff
diff --git a/libpod/runtime_pre_go1.20.go b/libpod/runtime_pre_go1.20.go
index 1c73043de..650e5e8d0 100644
--- a/libpod/runtime_pre_go1.20.go
+++ b/libpod/runtime_pre_go1.20.go
@@ -7,12 +7,11 @@ package libpod
 
 import (
 	"math/rand"
-	"time"
 )
 
 func init() {
 	// generateName calls namesgenerator.GetRandomName which the
 	// global RNG from math/rand. Seed it here to make sure we
 	// don't get the same name every time.
-	rand.Seed(time.Now().UnixNano())
+	rand.Seed(1024)
 }

```
```console
$ go version
go version go1.19.8 linux/amd64

$ for i in $(seq 1 10);do bin/podman run -d quay.io/libpod/testimage:20221018 top;done
...
$ bin/podman ps --format '{{.Names}}'|sort
blissful_feynman
cool_albattani
determined_colden
gifted_bhabha
gifted_thompson
magical_goldwasser
nifty_margulis
romantic_moore
sad_cartwright
wizardly_shaw
```

Comment (mtrmac) [reply test/e2e/run_test.go:51]: I’m sorry, I just meant that the thing we _want_ to test is generated names.

You’re right that this currently tests IDs.

---

> I can't get podman to reuse names even if I hardcode the seed:

Looking at the current `main`, there is also https://github.com/containers/podman/blob/0277adf69363da9bf6f92c5eb5c2e7d99a5e2446/vendor/github.com/containerd/stargz-snapshotter/estargz/testutil.go#L48-L50 (which, despite the `test` in the name, does, I think run in Podman production). That might be the cause, but I didn’t test in practice or debug further.

Comment (mtrmac) [reply test/e2e/run_test.go:51]: Separately,

> `for i in $(seq 1 10);do bin/podman run -d quay.io/libpod/testimage:20221018 top`

is it possible that older containers are still running while the loop proceeds to another iteration? (I didn’t test to confirm I’m afraid.)

In that case Podman would continue to try generating new names until there is no conflict (IOW, the non-deterministic seeding should improve performance to avoid retries, and make other corner cases rarer; it’s not because a conflict should immediately cause a failure).

Comment (cevich) [reply test/e2e/run_test.go:51]: Sorry for the confusion, the test I wrote is relying on the fact that podman will barf when trying to start another container with the same name as an existing one.  Collection of the CIDs is only done to help with cleanup.

Comment (edsantiago) [reply test/e2e/run_test.go:51]: Ahhhhh.... no, I got it now. Here, this should shed light:
```console
$ for i in 1 2 3;do bin/podman run -d quay.io/libpod/testimage:20221018 top >/dev/null; bin/podman inspect -l --format '{{.Name}}';done
gifted_bhabha
magical_goldwasser
gifted_thompson

! clean up
$ bin/podman rm -f -a >&/dev/null
$ !for
gifted_bhabha
magical_goldwasser
gifted_thompson
```
(that's just with my `1024` diff above. The `vendor` file is irrelevant).

So, there must be disambiguation code in podman name selection. That seems sensible. And we have confirmed that with a same seed, the **SEQUENCE** of names is what is reproducible. Finding a test for that is left as an exercise for the reader.

Comment (cevich) [reply test/e2e/run_test.go:51]: > continue to try generating new names until there is no conflict

Oooohhh, so generating duplicate names to existing containers is impossible then.  In that case the new test is useless.

Comment (cevich) [reply test/e2e/run_test.go]: okay, test is removed.

Comment (edsantiago) [reply test/e2e/run_test.go]: >a test could create a container, record the name, delete it; create another container, record the name, delete it. Then check that the two generated names are different

Exactly.
```console
$ for i in 1 2 3;do bin/podman run quay.io/libpod/testimage:20221018 true; bin/podman inspect -l --format '{{.Name}}';bin/podman rm -l &>/dev/null;done
gifted_bhabha
gifted_bhabha
gifted_bhabha
```

Comment (cevich) [reply test/e2e/run_test.go]: Okay ya, there you go.  That was with your fixed-seed though ya?   Hmmm, could be an worthwhile e2e test to have then?

Comment (edsantiago) [reply test/e2e/run_test.go]: Yes, with fixed-seed. I don't know enough to comment on how useful a test would be.

Comment (cevich) [reply test/e2e/run_test.go]: Well, IMHO under some use-case it seems it might be a fairly bad thing if podman started reusing names (or same sequence of names) like that.  The changes in this PR **DO** touch the RNG initialization code.  So perhaps it's worthwhile having a test.  Maybe let's do that as a followup PR though, this one is getting long in the tooth.

Comment (mtrmac) [reply test/e2e/run_test.go]: Ideally, all code should be reliable and robust against conflicts. I think that’s the case for the specific target of `generateNames` (or at least that one loop lined above), but it’s historically not the case for _all_ the random names the codebase generates (I think something in the link cluster around https://github.com/containers/storage/pull/1337 contains examples) — sometimes, on a conflict, we just fail.

And even if the loops do exist, they are rarely-exercised code that might fail, compare a recent https://github.com/containers/buildah/issues/4679 .

And the conflicts are more costly in performance (e.g. container creation time, which we have focused on recently) than not having conflicts.

So, on net, I would say that there is _some_ value to us having a test that flags if we ever regress to repeated name sequences. I don’t have much of an opinion on how much effort to spend on it, or, more importantly, if we can only cause a potentially-flaky test that depends on a RNG seed (with some ~2^-32?? probability), whether that flake rate increase is worth it.

**Comment by cevich** [comment]
This should be ready now.

**Review by mtrmac — COMMENTED** [review]
LGTM.

File: test/e2e/run_test.go (hunk 6 above)
Comment (cevich) [reply test/e2e/run_test.go]: Opened https://github.com/containers/podman/pull/18381
